### ISSUE TITLE
accept multiple sources to build compiler's context

### ISSUE THREAD

**Issue by lrlna** [apollographql/apollo-rs#221]
We will never really want to have a compiler that's created from a single graphql source, it's a lot more likely that the context is going to be built out from various sources (files, input strs etc). 

What does the API for this look like? What are some of use cases we should consider ahead of time?

**Comment by lrlna** [comment]
There are two different types of sources we should consider allowing to build the context with:
- graphql that contains **both schema + query**
- graphql that **only** contains **query**.

These should have separate API endpoints as well, since there are certain cases where you will want to make sure a query input does not modify your schema context. 

From the compiler's database point of view, we probably want to "warm start" the compiler with a single source, and then add additional sources, either full (graphql schema + query), or graphql query source, as inputs. The database should have query source as separate #[salsa::input] to avoid schema context manipulation when that's importatn. Because each input has its own key, we will be able to get the correct context when looking for it. 

Right now thinking this would be a rough diagram of a "warm start" with a schema, following by adding additional graphql query sources (input 1, input 2, input 3):
![Untitled_Artwork (2)](https://user-images.githubusercontent.com/8107784/184625589-07784dfc-cc87-4df7-bdd3-bb9648d7deaa.png)

**Comment by abernix** [comment]
Is it worth enumerating what I believe to be a third case:

* graphql that contains **just the schema**

In my experience, GraphQL that contains the schema and an operation is _uncommon_, but I'm certainly interested in the use cases.   I could actually use some help interpreting the specification here: the [Document](https://spec.graphql.org/draft/#sec-Document) section of the specification leads me to believe that a document with both seems to be deemed _not executable_ 🤔 .  There's [another section](https://spec.graphql.org/draft/#sec-Document-Syntax) of the specification with more detailed grammar, and it doesn't note that a document needs to be "one of" either [ExecutableDefinition](https://spec.graphql.org/draft/#ExecutableDefinition) or [TypeSystemDefinitionOrExtension](https://spec.graphql.org/draft/#TypeSystemDefinitionOrExtension) which means that they could be co-located with it, but maybe I'm not sure what the expectation of what a document would be.

**Comment by lrlna** [comment]
good third case addition!

there were at least 3 different people who've mentioned wanting the ability to parse both the executable (queries) and type system (schemas) definitions at the same time. I did not write down the reason for this, unfortunately. 

As for the spec, yes, you can have one or the other, and having a type system definition makes the document non-executable.

The compiler must be able to combine executable and type system definitions, however. That's how we get the nice type information, and validation for queries. The spec perhaps lends us nice naming for these compiler APIs:

```rust
// Creates compiler given an input source that has both executable
// and type system definition (for those who wanted this)
let compiler = ApolloCompiler::with_document(input);

// Creates compiler given a schema/type system definitions.
// Would be used as a warm start in the router, for example.
let compiler = ApolloCompiler::with_schema(input);

// Creates compiler with an executable definition.
// Question here: is there a world where you'd want to start a compiler
// with just the executable definition??
let compiler = ApolloCompiler::with_executable_definition(input);

// add additional executable_definition source
compiler.executable_definition(executable_input_1);
compiler.executable_definition(executable_input_2);
compiler.executable_definition(executable_input_3);


// add additional schema source
compiler.schema(schema_input_1);
compiler.schema(schema_input_2);
compiler.schema(schema_input_3);
```

**Comment by abernix** [comment]
Will `compiler.executable_definition` present any problems for the case where you want to build up a context with a _set_ of operations where, for example, each operation itself is an anonymous/unnamed operation and therefore invalid in a document, but valid on their own?  Could/should that be best represented as a set of compiler contexts each with their own executable definition?

**Comment by lrlna** [comment]
an anonymous operation is still valid. From the spec point of view it corresponds to `type Query` or whatever is defined as query in a schema definition, like in the example below an anonymous query will refer to `NewlyDefinedQuery`.

``` graphql
schema {
  query: NewlyDefinedQuery
}

type NewlyDefinedQuery {
  name: String
}
```

I think what you're trying to get at is in the case of the router, you might have separate anonymous operations, which all must be valid in the compiler. This is where creating a new salsa database for executable definitions will be useful ("Query Database" in the diagram above). Under the hood, i.e. not exposed to the user, each incoming anonymous operation definition will be treated as a separate input in the "Query Database" and probably return an ID we can then reference. We will then combine the executable definition input with the schema input in the "Schema Database", create any missing relationships, validate and have all of that be accessible via the "Root Database". We should then be able to access whichever combined context of schema + executable definition via the ID we returned earlier.

**Comment by abernix** [comment]
Cool, I think that could work!  I'm interested in any other thoughts, too.

And yes, you've caught onto my important nuance that I could have been even more clear about: I was trying to dig into whether we would run into any constraint where "more than one anonymous operation existed in a context" but I was thinking about it as a flat-namespace where it would encounter the "no more than one anonymous operation" case, but it doesn't sound like that need to be a concern with the executable database. 

Thanks!

**Comment by lennyburdette** [comment]
All I'm looking for is something like this: 

```rs
struct MyPlugin {
    configuration: Conf,
    schema: Arc<ApolloCompiler>,
}

#[async_trait::async_trait]
impl Plugin for MyPlugin {
    type Config = Conf;

    async fn new(init: PluginInit<Self::Config>) -> Result<Self, BoxError> {
        Ok(MyPlugin {
            configuration: init.config,
            schema: init.supergraph_schema,
        })
    }

    fn router_service(
        &self,
        service: BoxService<RouterRequest, RouterResponse, BoxError>,
    ) -> BoxService<RouterRequest, RouterResponse, BoxError> {
        let schema = self.schema.clone();

        ServiceBuilder::new()
            .checkpoint(move |req: RouterRequest| {
                let ctx = schema.add_input(req.originating_request.body().query);
                //               ^^^^^^^^^ – no idea what to call this!

                let operation =
                    ctx.operation_by_name(req.originating_request.body().operation_name);

                // analyze operation and maybe reject request
            })
            .service(service)
            .boxed()
    }
}
```

**Comment by lrlna** [comment]
@lennyburdette something of the sort will work - that's the scenario of "start with a schema and then add queries".

**Comment by lrlna** [comment]
closed in #277
